driver/src/main/com/mongodb/GridFSBucketImpl.java

```diff
@@ -0,0 +1,316 @@
+/*
+ * Copyright 2015 MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb;
+
+import com.mongodb.client.GridFSBucket;
+import com.mongodb.client.GridFSDownloadStream;
+import com.mongodb.client.GridFSFindIterable;
+import com.mongodb.client.GridFSUploadStream;
+import com.mongodb.client.MongoCollection;
+import com.mongodb.client.MongoDatabase;
+import com.mongodb.client.model.GridFSDownloadByNameOptions;
+import com.mongodb.client.model.GridFSUploadOptions;
+import com.mongodb.client.model.IndexOptions;
+import com.mongodb.client.model.Indexes;
+import com.mongodb.client.result.DeleteResult;
+import com.mongodb.client.result.UpdateResult;
+import org.bson.BsonDocument;
+import org.bson.BsonInt32;
+import org.bson.BsonObjectId;
+import org.bson.BsonString;
+import org.bson.Document;
+import org.bson.codecs.configuration.CodecRegistry;
+import org.bson.conversions.Bson;
+import org.bson.types.ObjectId;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+
+import static com.mongodb.ReadPreference.primary;
+import static com.mongodb.assertions.Assertions.notNull;
+import static java.lang.String.format;
+
+final class GridFSBucketImpl implements GridFSBucket {
+    private final MongoDatabase database;
+    private final String bucketName;
+    private final int chunkSizeBytes;
+    private final CodecRegistry codecRegistry;
+    private final WriteConcern writeConcern;
+    private final ReadPreference readPreference;
+    private final MongoCollection<BsonDocument> filesCollection;
+    private final MongoCollection<BsonDocument> chunksCollection;
+    private boolean checkedIndexes;
+
+    GridFSBucketImpl(final MongoDatabase database) {
+        this(database, "fs");
+    }
+
+    GridFSBucketImpl(final MongoDatabase database, final String bucketName) {
+        this.database = notNull("database", database);
+        this.bucketName = notNull("bucketName", bucketName);
+        this.chunkSizeBytes = 255;
+        this.codecRegistry = database.getCodecRegistry();
```

Comment:
This class should have its own codec registry that does exactly what is required by the GridFS application